[GlobalISel][AArch64] Add libcall lowering for fpowi.

davemgreen opened the pull request:
This adds legalization, notably libcall lowering for fpowi. It is a little different to other methods as the function takes both a float and integer register. Otherwise all vectors get scalarized and fp16 is promoted to fp32.

aemerson commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -851,6 +853,24 @@ LegalizerHelper::libcall(MachineInstr &MI, LostDebugLocObserver &LocObserver) {
+  case TargetOpcode::G_FPOWI: {
+    Type *HLTy = getFloatTypeForLLT(Ctx, LLTy);
+    Type *ITy = IntegerType::get(
+        Ctx, MRI.getType(MI.getOperand(2).getReg()).getSizeInBits());
```
You can `cast<GenericMachineInstr>(MI)` and then use `getReg(N)` for brevity here and below as well.

aemerson commented on llvm/test/CodeGen/AArch64/fpowi.ll:
```diff
@@ -0,0 +1,1422 @@
+; CHECK-GI-NEXT:    str x30, [sp, #-16]! // 8-byte Folded Spill
+; CHECK-GI-NEXT:    .cfi_def_cfa_offset 16
+; CHECK-GI-NEXT:    .cfi_offset w30, -16
+; CHECK-GI-NEXT:    bl __powidf2
```
Ideally we should be tail-calling these. We already can do this for memcpy libcalls, could you add support in `createLibCall()`? I think you might just need to copy:
```
  Info.IsTailCall = MI.getOperand(MI.getNumOperands() - 1).getImm() &&
                    isLibCallInTailPosition(MI, MIRBuilder.getTII(), MRI);
```
from `createMemLibcall()`

davemgreen replied:
That sounds good, but it should probably be a separate patch as it is a little more involved.

davemgreen commented:
^ Rebased.

I was looking at libcall lowering a little while ago, but the patch I had was currently miscompiling when the return types did not match.

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -979,10 +982,30 @@ LegalizerHelper::libcall(MachineInstr &MI, LostDebugLocObserver &LocObserver) {
+      return UnableToLegalize;
+    }
+    auto Libcall = getRTLibDesc(MI.getOpcode(), Size);
+    SmallVector<CallLowering::ArgInfo, 2> Args;
+    Args.push_back({GMI.getReg(1), HLTy, 0});
+    Args.push_back({GMI.getReg(2), ITy, 1});
```
Initializer list

aemerson commented:
Did your recent work on tail calls allow us to tail call here now?

davemgreen committed:
[GlobalISel][AArch64] Add libcall lowering for fpowi.

This adds legalization, notably libcall lowering for fpowi. It is a little
different to other methods as the function takes both a float and integer
register. Otherwise all vectors get scalarized.

davemgreen commented:
Yeah I was just rebasing!

kiranchandramohan commented:
@davemgreen The patch seems to have some issues with the Windows target. The dump after the legalizer prints BL without symbol as given below.
```
BL &"", <regmask $fp $lr $wzr $xzr $b8 $b9 $b10 $b11 $b12 $b13 $b14 $b15 $d8 $d9 $d10 $d11 $d12 $d13 $d14 $d15 $h8 $h9 $h10 $h11 $h12 $h13 $h14 $h15 $s8 $s9 $s10 $s11 $s12 and 55 more...>, implicit-def $lr, implicit $sp, implicit $d0, implicit $w0, implicit-def $d0
``

It works fine on the linux target and prints (BL &__powidf2 .....).

Could you have a look?

Reproducer:
```
; ModuleID = 'LLVMDialectModule'
source_filename = "LLVMDialectModule"
target datalayout = "e-m:w-p:64:64-i32:32-i64:64-i128:128-n32:64-S128"
target triple = "aarch64-pc-windows-msvc19.37.32822"

@_QFpf07aaECtwo = internal constant double 2.000000e+00
@_QFpf07aaECwp = internal constant i32 8

define void @pf07aa_() {
  %1 = alloca i32, i64 1, align 4
  %2 = alloca double, i64 1, align 8
  %3 = load i32, ptr %1, align 4
  %4 = call contract double @llvm.powi.f64.i32(double 2.000000e+00, i32 %3)
  store double %4, ptr %2, align 8
  ret void
}

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare double @llvm.powi.f64.i32(double, i32) #1

attributes #0 = { nocallback nofree nosync nounwind willreturn }
attributes #1 = { nocallback nofree nosync nounwind speculatable willreturn memory(none)
```

davemgreen commented:
Hi - yeah will do, I'll take a look. It sounds like the libcall is missing, and perhaps should be failing to legalize.
